Firestore/CMakeLists.txt

```diff
@@ -15,6 +15,53 @@
 cmake_minimum_required(VERSION 2.8.11)
 project(firestore C CXX)
 
+option(ASAN "Build with Address sanitizer (mutually exculsive with TSAN)" OFF)
+# TODO(varconst): msan
+# Memory sanitizer is more complicated:
+# - it requires all dependencies to be compiled with msan enabled (see
+#   https://github.com/google/sanitizers/wiki/MemorySanitizerLibcxxHowTo);
+# - AppleClang doesn't support it.
+option(TSAN "Build with Thread sanitizer (mutually exculsive with ASAN)" OFF)
+option(UBSAN "Build with Undefined behavior sanitizer" OFF)
+
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  set(HAVE_SANITIZERS 1)
+endif()
+if(ASAN OR TSAN OR UBSAN)
+  set(USE_SANITIZERS 1)
+endif()
+
+if (NOT HAVE_SANITIZERS AND USE_SANITIZERS)
+  message(FATAL_ERROR "Only Clang and GCC support sanitizers.")
+endif()
+
+function(add_to_compile_and_link_flags flag)
```

Comment:
If you make this a macro instead of a function you can avoid the `PARENT_SCOPE` messiness.

Reply:
Done.